File: DFS/Cpp/1012.cpp
// 유기농 배추
#include <iostream>
#include <stack>
#include <vector>
using namespace std;

int tile[50][50];
int check[50][50];
int count;
int dx[] = {-1, 0, 0, 1};
int dy[] = {0, -1, 1, 0};

void dfs(int a, int b) {
    stack<pair<int, int>> s;
    s.push(make_pair(a, b));
    while (!s.empty()) {
        int x, y;
        pair<int, int> p = s.top();
        s.pop();
        x = p.first;
        y = p.second;
        check[x][y] = 1;
        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            if (nx < 0 || nx > 49 || ny < 0 || ny > 49) continue;
            if (check[nx][ny] == 0 && tile[nx][ny] == 1)
                s.push(make_pair(nx, ny));
        }
    }
}

int main() {
    int testCase, x, y, n;
    cin >> testCase;
    for (int i = 0; i < testCase; i++) {
        count = 0;
        for (int j = 0; j < 50; j++) {
            for (int k = 0; k < 50; k++) {
                tile[j][k] = 0;
                check[j][k] = 0;
            }
        }
        cin >> x >> y >> n;
        for (int j = 0; j < n; j++) {
            int a, b;
            cin >> a >> b;
            tile[a][b] = 1;
        }
        for (int j = 0; j < 50; j++) {
            for (int k = 0; k < 50; k++) {
                if (check[j][k] == 0 && tile[j][k] == 1) {
                    count++;
                    dfs(j, k);
                }
            }
        }
        cout << count << '\n';
    }
}

File: DFS/Python/1012.py
# 유기농 배추

dx = [-1, 0, 0, 1]
dy = [0, -1, 1, 0]


def dfs(tile, check, x, y):
    global dx
    global dy
    stack = []
    stack.append([x, y])
    while stack:
        x, y = stack.pop()
        check[x][y] = 1
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if nx < 0 or nx > 49 or ny < 0 or ny > 49:
                continue
            if check[nx][ny] == 0 and tile[nx][ny] == 1:
                stack.append([nx, ny])


testCase = int(input())
for _ in range(testCase):
    count = 0
    tile = list([0 for _ in range(50)] for _ in range(50))
    check = list([0 for _ in range(50)] for _ in range(50))
    M, N, n = list(map(int, input().split()))
    for _ in range(n):
        a, b = list(map(int, input().split()))
        tile[a][b] = 1
    for i in range(50):
        for j in range(50):
            if check[i][j] == 0 and tile[i][j] == 1:
                count += 1
                dfs(tile, check, i, j)
    print(count)
